src/shell/shell.cpp

```diff
@@ -240,9 +240,21 @@ Bitu DOS_Shell::GetRedirection(char *s, char **ifn, char **ofn, bool *append)
 			*ifn = temp;
 			continue;
 
-		case '|': ch = 0; num++;
+		case '|':
+			num++;
+			if (*pipe)
+				free(*pipe);
+			lr = ltrim(lr);
+			*pipe = lr;
+			while (*lr)
+				lr++;
+			temp_len = static_cast<size_t>(lr - *pipe + 1u);
+			temp = new char[temp_len];
+			safe_strncpy(temp, *pipe, lr - *pipe + 1);
+			*pipe = temp;
+			continue;
 		}
-		*lw++=ch;
+		*lw++ = ch;
```

Comment:
This function (GetRedirection) is actually the headliner example for technical debt in ticket https://github.com/dosbox-staging/dosbox-staging/issues/1314, which calls for refactoring away from this function's historic coding style of using heavily ultra-terse, heavily nested, poorly documented, and pointer-laden code.

Reply:
So although this commit (as written) adds more of the same style of code flagged in #1314, the good news is this PR can start to turn this function around and begin to address the issues flagged in #1314 by adding using tests around this pipe feature and re-writing it using C++17 features mentioned in #1314.

Reply:
The flagged block of code has been improved to this:
``` c++
	// Get the current content of the redirection
	line_read = ltrim(line_read);
	redir_str = *redir = line_read;
	// Try to find the characters for string split
	find_chars = character == '|'
	                     ? ""
	                     : (character != '<' ? " |<" : " |>");
	found = redir_str.find_first_of(find_chars);
	// Get the length of the substring before the
	// characters, or the entire string if not found
	if (found == std::string::npos)
		temp_len = redir_str.size();
	else
		temp_len = found -  // Ignore ':' character
		           (redir_str[found - 1] == ':' ? 1 : 0);
	// Assign substring content of the length to output parameters
	line_read += temp_len;
	redir_str[temp_len] = 0;
	temp = new char[temp_len + 1];
	safe_strncpy(temp, redir_str.c_str(), temp_len + 1);
	*redir = temp;
	continue;
}
*line_write++ = character;
```

It now has comments, the two-letter variables have been expanded, and some logic has been made less terse. But I still found it hard to grasp, so to fix the memory leak I had to follow the rote logic but without fully understanding it.

The main problem is the arguments into this function are double-pointers, for which this section `delete`s or `new`s memory, that subsequently has to be managed externally.  The use of double-pointers, "temp" and "temp_str" variables, and many substring calls and offset management makes it hard to see the bigger picture of what it's trying to accomplish through all of the fiddly details.

I think we should flag this area with a FIXME comment to refactor it along with its calling function to do away with the double pointers and low-level C-string management.

Reply:
I fully agree that the existing code was hard to read and understand. I refactored the code again to use std::string instead of double pointers so that hopefully the code will be easier to understand than before. Thanks for pushing a commit meant to fix the said leak flagged by the sanitizer. Meanwhile, the use of "new char" is completely removed in the updated code, which may help avoiding the leak as well.